scipy/signal/signaltools.py

```diff
@@ -463,6 +460,9 @@ def convolve(in1, in2, mode='full'):
     volume = asarray(in1)
     kernel = asarray(in2)
 
+    if not _is_valid_mode_shapes(mode, volume.shape, kernel.shape):
+        volume, kernel = kernel, volume
```

Comment:
Your implementation in your comment is slightly buggy AFAICT.  What happens if the `shape2[0] == shape1[0]` but `shape2[i] > shape1[i]` for the remaining `i`?